pisa/stages/xsec/dis_sys.py

```diff
@@ -0,0 +1,213 @@
+            output_apply_keys=output_apply_keys,
+        )
+
+        assert self.input_mode is not None
+        assert self.calc_mode is None
+        assert self.output_mode is not None
+
+        self.extrapolation_type = extrapolation_type
+
+    @profile
+    def setup_function(self):
+
+        extrap_dict = from_file('cross_sections/tot_xsec_corr_Q2min1_isoscalar.pckl')
+
+        # load splines
+        wf_nucc =    from_file('cross_sections/dis_csms_splines_flat/NuMu_CC_flat.pckl')
+        wf_nubarcc = from_file('cross_sections/dis_csms_splines_flat/NuMu_Bar_CC_flat.pckl')
+        wf_nunc =    from_file('cross_sections/dis_csms_splines_flat/NuMu_NC_flat.pckl')
+        wf_nubarnc = from_file('cross_sections/dis_csms_splines_flat/NuMu_Bar_NC_flat.pckl')
+
+        # set this to events mode, as we need the per-event info to calculate these weights
+        self.data.data_specs = 'events'
+
+        for container in self.data:
+
+            # creat keys for external dict
+
+            if container.name.endswith('_cc'):
+                current = 'CC'
+            elif container.name.endswith('_nc'):
+                current = 'NC'
+            else:
+                raise ValueError('Can not determine whether container with name "%s" is pf type CC or NC based on its name'%container.name)
+            nu = 'Nu' if container['nubar'] > 0 else 'NuBar'
+
+            lgE = np.log10(container['true_energy'].get('host'))
+            bjorken_y = container['bjorken_y'].get('host')
+            dis = container['dis'].get('host')
+
+            #
+            # Calculate variation of total cross section
+            #
+
+            lgE_min = 2. if self.extrapolation_type == 'constant' else 1.68
+            mask = lgE <= lgE_min
+
+            w_tot = np.ones_like(lgE)
+            
+            poly_coef = extrap_dict[nu][current]['poly_coef']
+            w_tot[~mask] = np.polyval(poly_coef, lgE[~mask])
+           
+            if self.extrapolation_type == 'constant':
+                poly_coef = extrap_dict[nu][current]['poly_coef']
+                w_tot[mask] = np.polyval(poly_coef, lgE_min * w_tot[mask]) 
+            elif self.extrapolation_type == 'linear':
+                lin_coef = extrap_dict[nu][current]['linear']
+                w_tot[mask] = np.polyval(lin_coef, lgE[mask])
+            elif self.extrapolation_type == 'higher':
+                poly_coef = extrap_dict[nu][current]['poly_coef']
+                w_tot[mask] = np.polyval(poly_coef, lgE[mask]) 
+            else:
+                raise ValueError('Unknown extrapolation type "%s"'%self.extrapolation_type)
+
+            # mke centered arround 0, and set to 0 for all non-DIS events
+            w_tot -= 1
+            w_tot *= dis
+          
+            container["dis_correction_total"] = w_tot
+
+            #
+            # Calculate variation of differential cross section
+            #
+
+            lgE_min = 2.
+            mask = lgE <= lgE_min
```

Comment:
again, use a meaningful mask name, and also again, not clear why the endpoint isn't included in valid range? Also, need both masks multiple times, so avoid multiple negations of the entire array by defining both:
E.g.,
```python
spline_valid_mask = lgE >= lgE_min
spline_invalid_mask = ~spline_valid_mask
```